code/modules/power/turbine/turbine.dm

```diff
@@ -0,0 +1,557 @@
+	turbine_mixture = null
+	all_parts_connected = FALSE
+	SSair.stop_processing_machine(src)
+
+/obj/machinery/power/turbine/core_rotor/attackby(obj/item/object, mob/user, params)
+	if(!panel_open)
+		balloon_alert(user, "open the maintenance hatch first")
+		return ..()
+	if(all_parts_connected)
+		if(istype(object, compressor.part_path))
+			var/obj/item/turbine_parts/compressor/compressor_part = object
+			if(!compressor.installed_part)
+				if(!do_after(user, 2 SECONDS, src))
+					return
+				user.transferItemToLoc(compressor_part, src)
+				compressor.installed_part = compressor_part
+				compressor_part_efficiency = compressor_part.part_efficiency
+				calculate_parts_limits()
+				balloon_alert(user, "installed new part")
+				return
+			if(compressor.installed_part.part_efficiency < compressor_part.part_efficiency)
+				if(!do_after(user, 2 SECONDS, src))
+					return
+				user.transferItemToLoc(compressor_part, src)
+				user.put_in_hands(compressor.installed_part)
+				compressor.installed_part = compressor_part
+				compressor_part_efficiency = compressor_part.part_efficiency
+				calculate_parts_limits()
+				balloon_alert(user, "replaced part with a better one")
+				return
+
+			balloon_alert(user, "already installed")
+			return
+
+		if(istype(object, turbine.part_path))
+			var/obj/item/turbine_parts/stator/stator_part = object
+			if(!turbine.installed_part)
+				if(!do_after(user, 2 SECONDS, src))
+					return
+				user.transferItemToLoc(stator_part, src)
+				turbine.installed_part = stator_part
+				stator_part_efficiency = stator_part.part_efficiency
+				calculate_parts_limits()
+				balloon_alert(user, "installed new part")
+				return
+			if(turbine.installed_part.part_efficiency < stator_part.part_efficiency)
+				if(!do_after(user, 2 SECONDS, src))
+					return
+				user.transferItemToLoc(stator_part, src)
+				user.put_in_hands(turbine.installed_part)
+				turbine.installed_part = stator_part
+				stator_part_efficiency = stator_part.part_efficiency
+				calculate_parts_limits()
+				balloon_alert(user, "replaced part with a better one")
+				return
+
+			balloon_alert(user, "already installed")
+			return
+
+	if(istype(object, part_path))
+		var/obj/item/turbine_parts/rotor/rotor_part = object
+		if(!installed_part)
+			if(!do_after(user, 2 SECONDS, src))
+				return
+			user.transferItemToLoc(rotor_part, src)
+			installed_part = rotor_part
+			rotor_part_efficiency = rotor_part.part_efficiency
+			calculate_parts_limits()
+			balloon_alert(user, "installed new part")
+			return
+		if(installed_part.part_efficiency < rotor_part.part_efficiency)
+			if(!do_after(user, 2 SECONDS, src))
+				return
+			user.transferItemToLoc(rotor_part, src)
+			user.put_in_hands(installed_part)
+			installed_part = rotor_part
+			rotor_part_efficiency = rotor_part.part_efficiency
+			calculate_parts_limits()
+			balloon_alert(user, "replaced part with a better one")
+			return
+
+		balloon_alert(user, "already installed")
```

Comment:
This seems like logic that should be handled on one parent proc that all these parts share, rather then being copy pasted in attackby like this

Reply:
Yeah I know, I was already planning on cleaning it up